### ISSUE TITLE
AGS 4: support Alpha in Color parameter

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#2525]
This is written after #1514 and #1980.

Now when AGS supports true 32-bit colors for drawing in game, we may consider supporting Alpha value as well.
In terms of storing Alpha component in a color number, this is already supported by the field format, but there are several issues to resolve before this may be actively used.

The subtasks follow:

- [x] 1. Expand Game.GetColorFromRGB to also have "alpha" parameter. For backwards compatibility, and also because it's less common, add it as a last optional parameter with default value of 255 (full alpha).
EDIT: alternatively, make another function for getting ARGB/RGBA.
- [x] 2. Make sure that all the drawing operations in Allegro support alpha blending (not just blit, bit also draw line etc). I've never investigated that, but this may not be trivial, depending on how they work. **IF** Allegro does not support that, then we might have to implement our own, by cloning standard allegro functions, and more or less passing any pixel assignment in them through a blender (pixel function). These functions will be called if alpha if anything but 0 or 255.
In case alpha is 255 the standard functions are called;
In case alpha is 0, nothing is drawn and whole operation may be skipped.
- [x] 3. If alpha is supported, then all the object Color properties must respect alpha component as well. Such as: GUI colors, speech color, and so forth.
- [x] 4. BTW, there will be a side effect of this, that color 0 which is currently used in some cases meaning "transparent" (even in 32-bit games) will represent a 0,0,0,0 ARGB (so color with alpha == 0), and therefore it will be possible to assign true-black color to GUI background and such using any non-zero alpha. But this has to be double checked.
- [ ] 5. Editor: there's a issue of the color picker. The common color picker dialog which we currently use does not let input alpha. Maybe there's a mode in which it does, then it will simply have to be enabled. But if not, then we'll require a different dialog.
In short term we may leave things as is, but then you won't be able to set alpha in the editor other than by editing a Color value by hand (and choosing a different color in the picker dialog will reset alpha to 255).
EDIT: split into ticket #2656
- [x] 6. Editor: a minor issue is that since all color properties are `ints`, that is - signed ints, having alpha in them will make them go negative. That's not an issue in UI anymore, since I changed all Color properties to be displayed as RGB sequence, but they will be written as negative in the Game.agf xml. What can be done about it? We may change all of these properties to `uint`; but these properties are in AGS.Types, so I wonder if that will affect editor plugins. I am not versed enough in how C# interfaces work to predict the consequence of this change. Another way is to serializes these as hex. We'd probably need to add a new attribute "SerializeAsHex" for these properties.

Is there anything else that comes to mind that might require attention?

**UPDATE:**
Subtasks 4 and 6 are resolved by #2648
Subtask 2 is resolved within #2661

**Comment by ericoporto** [comment]
Uhm, the text functions use blit so if the color alpha isn't 0 I think they are going to be weird. 

Other issue may be anything that has color and transparency.

**Comment by ivan-mogilko** [comment]
Having alpha in color is the same as having alpha in sprites. All the logical and technical difficulties are shared between these 2 cases.

**Comment by ericoporto** [comment]
my main concern : https://github.com/adventuregamestudio/ags/blob/6ce71b446cb7a561dc4fbcec57f71a29547c5b1e/Engine/gfx/gfxdriverbase.cpp#L496-L518

this is_mask check is about to be more expensive, this is the most expensive function in the bitmap to video memory.

Other than this, all the pixelperfect tests that currently simply compare with the mask color.

plus blitting text where the color is alpha is going to puncture holes in the dialog boxes.

**Comment by ivan-mogilko** [comment]
That alpha processing replaces colors with alpha == 0 with a mask color (00FF00FF in 32-bit). It does not affect those pixels with non-zero alpha, they will keep their alpha value.

What I meant is, AGS already may have sprites with pixels where alpha is between 0 and 255, which come by importing sprites. These sprites already affect pixel-perfect detection logic, for example. Having those alpha values put by a drawing operation in script will not add anything new here.

Similarly, these sprites with varied alpha already pass through bitmap->texture conversion. Having pixels drawn with alpha won't add anything new here either.

The only thing which will affect logic here are colors with 0 alpha. As you rightly mentioned, right now they are replaced by the "transparent color" constant. I did not yet decide what should happen with the "transparent color" constant, this is something to think well about, keeping consequences in mind. But in the worst case, in a short term, we may keep the rule that any color with alpha 0 becomes a "mask color". Sure that will be strange (to read these pixels back), but will keep things consistent so long as this post-processing exists in sprite load.

> plus blitting text where the color is alpha is going to puncture holes in the dialog boxes - they would need to blit to make the border in a separate bitmap and then later blended.

Yes, I am aware of that. That's why I mentioned that possibly we'll have to have 2 sets of functions: for opaque drawing color, and non-opaque drawing color, where the first uses fast blitting and second uses an extra processing. This way common operations will be kept as fast as before.

> sure you need to do something like FFacolor now because you need the alpha to be 255. this means we are always using a negative int.

That is correct. Do you see a problem with that?

**Comment by ivan-mogilko** [comment]
> my main concern :
> this is_mask check is about to be more expensive, this is the most expensive function in the bitmap to video memory.

Could you elaborate, why would it be more expensive, and in which circumstances?
I did not think about this particular function yet, but in a hypothetical case when the "mask color" is gone, the equality check
```return c == MASK_COLOR_32```
will likely be replaced by a bitwise AND, something like:
```return static_cast<bool>(c & 0xFF000000)```

**Comment by ericoporto** [comment]
I think overall the idea of keeping the mask color is a good one, simply because most image drawing software do not keep the colors when drawing with zero alpha and write with 0 instead to all colors

> we may keep the rule that any drawing color with alpha 0 becomes a "mask color". Sure that will be strange (to read these pixels back),

I believe this will save us from debugging lots of user images where their software is not saving what is in memory correctly. Plus if we are going to have to do the fix for linear rendering anyway for our own drawing surfaces, there isn't much to gain from importing the images from users and keeping their alpha untouched (supposing someone uses a software that allows to correctly prepare the 0 alpha colors correctly for linear rendering).

**Comment by AlanDrake** [comment]
1. `Game.GetColorFromRGB` should stay for RGB, instead we should create a new `Game.GetColorFromRGBA` which is cleared in its intent.

5. The native color picker doesn't have alpha support. We'll need a custom component.

6. Color values being negative should not be a problem. Storing them in hexadecimal format is probably better, if there are no major obstacles in doing so.

As for scripting, what is going to happen with `.XXXColor` properties?
- Do we break compat and use the real value? Ancillary methods like `.XXXColorRGB/.XXXColor24` could be added by a module for convenience. I don't think it would be too much work to upgrade a project.
- Will they remain as 24bit for compatibility, and perhaps add a `.XXXColorARGB` that exposes the full value?
- Maybe a general setting to enable/disable ARGB behavior? Though, that would be annoying to check on every place, and it would litter the code for the sake of backward compat, which ags4 is meant to remove.

**Comment by ivan-mogilko** [comment]
> As for scripting, what is going to happen with .XXXColor properties?

My first thought was that these will need to be updated on older project / game load, but that is it. The colors should not be set directly at all, but using Game.GetColorFromRGB(A), which solves the compatibility issues in script (I guess?).
Do you see any possible troubles with that?

**Comment by ericoporto** [comment]
> The colors should not be set directly at all, but using Game.GetColorFromRGB(A)

Why not setting them directly? Couldn't they be set with hex like 0xFFRRGGBB ?

By using Game.GetColorFromRGB it's yet another API call - so even slower Get/SetPixel and all other drawing methods, plus all the drawing stuff becomes slower.

**Comment by ivan-mogilko** [comment]
> Why not setting them directly? Couldn't they be set with hex like 0xFFRRGGBB ?

I mean, for the most of users it would be convenient to use GetColorFromRGB, instead of inserting color numbers received from "Color finder" for example. Of course if you know how to encode RGB, then you might as well do that.

**Comment by ivan-mogilko** [comment]
> Why not setting them directly? Couldn't they be set with hex like 0xFFRRGGBB ?

Mmm, yes, that is true.

For the most of the common users, and simple cases, GetColorFromRGB would be more convenient, as that does not require to worry about how values are formed.

Setting encoded values directly will be faster, but unfortunately requires to comply to the current form.
So, unfortunately, if using these values, they, or their generation in script, will have to be adjusted depending on the current format.

**Comment by AlanDrake** [comment]
But AGS4 is meant to drop retrocompat. I'd rather we did not keep special cases, unless for the most dire necessity.

Even in the worst case, a user would spend no more than a couple of days fixing their color assignment instances.

Would it be possible to have alternative assignments like erico suggested? Like `obj.Color.RGB = 0xRRGGBB` while still letting `obj.Color = 0xFFRRGGBB` work for direct color?

**Comment by ivan-mogilko** [comment]
Any kind of multiple assignment would require to replace `int XColor` properties with a struct that has multiple properties. Along with the change of API, that will make such assignment slower (I mention this as performance concern was noted above).

Another approach could be in supporting parameterized macros, in which case we could replace GetColorFromRGB with a macro too, and have macros like GetOpaqueColor(r,g,b) which would add 0xFF alpha.

But I don't have a defined opinion on this atm.

There's an alternative posted as #1514 originally, where Alpha is a standalone property for drawing color in DrawingSurface.

**Comment by ivan-mogilko** [comment]
I see now that I opened this ticket too early, without thinking it all through from the design perspective.

There are cases where having alpha in a color will either duplicate existing feature, or may not be enough to achieve all wanted effects.

1. For objects that may have either a color or a graphic, such as GUI, there's already Transparency property that does exactly what alpha component would do in a Color property. Which means that use of alpha in the Color there is mostly useless (at least without changing the meaning of existing properties).

2. For DrawingSurface, there are 2 separate operations that may be wanted:
* write alpha into surface pixels (copy color)
* draw something with alpha (blend color)
Unfortunately, I did not ask @fernewelten which exactly effect of these two did he have in mind when writing #1514.
But having an alpha component in DrawingColor will likely achieve the second effect, while the first effect will still be not possible to achieve. For the first effect we'd need a separate drawing parameter, something that selects between Copying pixels and Blending pixels.

**Comment by ivan-mogilko** [comment]
In regards to approaches suggested in the earlier conversation here, there are 3 distinct proposals. Few days ago we [talked about this again on AGS Forums](https://www.adventuregamestudio.co.uk/forums/index.php?msg=636668384), and I will copy parts of my comments here, as a summary to these.

---

**Approach 1: Color as a managed struct**

Where you access colors as `object.Color.RGB = opaque_color` or `object.Color.ARGB = color_with_alpha`, and similar.
In short, I like this as a syntax, but don't like the amount of effort it will require to make something like that work in AGS (or rather effort / benefit ratio).

There are multiple issues here: the color storage in dynamic memory that has to be serialized, how game objects will be accessing these, will the Color instance duplicate color values from object's own data or read/write directly to them, etc. There will be a use of Color in object properties, but also in user script, then likely it has to contain all data within itself. AGS engine's code currently has a number of problems in its organization, which may make it inconvenient to bind a game class to a separate managed object. An example is GUI classes that do not have any proper "runtime" class equivalent yet which would allow to handle access to managed data. We'd need to reorganize them first prior to doing something like that.

If we replace types of properties, then should likely also replace function arguments. There is "color" arg in few functions, like Overlay.CreateTextual.

If Color is a managed struct, then it's assigned by a reference and shared, meaning that changing it in one reference will change all of instances. We may try making builtin object properties make a full copy instead of assigning a Color reference, but user variables in script will still be references. It maybe would be better if Color were a non-referenced POD type (plain struct), but then we need AGS script to implement plain struct assignment by value.

So, such approach has multiple prerequisites, but doing all this just to be able to assign "alpha" value to a color sometimes seems unjustified at this point in engine development.

**Approach 2: Have all colors treated uniformly as ARGB in all properties and arguments.**

That will be most consistent, and simple to implement, but will make 0xFF... prefix mandatory for opaque colors when you define the color by hand, which may be seen as annoying by some, and be prone to user mistakes.
Two more notes here:
- Value 0 will be the easy way to set "transparent color".
- Alpha has to be in highest bits, because 0xRRGGBB may be cast (assigned) to 0xAARRGGBB, but cannot be cast to 0xRRGGBBAA (will require value conversion).
- Majority of users can just use Game.GetColorFromRGB / Game.GetColorFromRGBA functions to form colors; where old one will always return opaque color (alpha 0xFF).

**Approach 3: Have separate properties for RGB and ARGB.**

After some hesitation, I suppose that this is also a viable alternative.
This means that for each object.xColor property there's object.xColorARGB (or I may suggest shorter `xColorA` variant).
The reasoning here is that in majority of cases users will likely want opaque colors, and specify alpha in relatively very small number of situations. So for the most part they will be using object.xColor as usual, but when they need alpha, they use object.xColorA.

In this pair Object.Color would interpret positive integer values as RGB, and let's say "-1", or any invalid value would meaning "transparent color".
And Object.ColorA would interpret any value as ARGB.

The upsides of this approach are:
- This allows to restrict certain things to no alpha support, and make alpha support more explicit.
- Users are not required to remember to put alpha 0xFF when they define opaque colors.

The downsides:
- Obvious duplication of each Color property (but I think there are not too many of those).
- There is also a room for mistake, where one would try to assign full argb to rgb property.
- There's a remaining question of colors passed as function parameters, which I could not solve.

**Comment by messengerbag** [comment]
I think I somewhat naively assumed that supporting proper 32-bit color meant that all colors would be 32-bit. So my default assumption was Approach 2.

I don't think the issue about people constructing color values manually is very serious. Anyone who is doing that will be able to adapt. For everyone else, a robust and convenient API (e.g. to adjust the opacity of a given color, or get its fully-opaque version) is key.

Alternative 1 is the one I think would be most elegant from a user POV, _if_ the building blocks were there. I feel it ought to be a non-managed struct with only one field, but you'd need to be able to pass it to as an argument. It's probably not so much better that it justifies the work.

For Alternative 3, I'm mainly concerned about functions that return a color (like `DrawingSurface.GetPixel`): will there now need to be multiple versions of those as well? I think it will be challenging for users to keep track of whether their variables represent 32-bit or 24-bit colors. I am not in favor of this option.

Aside from that, I would also suggest replacing `COLOR_TRANSPARENT` with `Colors.Transparent`, a pseudo-enum that would allow a list of basic colors: Colors.Black, .White, .Red, .Green, .Blue, .Magenta, .Yellow, .Aqua, etc.

For example, I would suggest the API in action might look something like:

```  mySurface.DrawingColor = Color.SetOpacity(Colors.Red, 128); // 50% opaque red
  mySurface.DrawPixel(10,10);  // Mixed with whatever the underlying color is
  int mixColorOpaque = Color.Opaque(mySurface.GetPixel(10,10));```

**Comment by ivan-mogilko** [comment]
> I'm mainly concerned about functions that return a color (like DrawingSurface.GetPixel): will there now need to be multiple versions of those as well? I think it will be challenging for users to keep track of whether their variables represent 32-bit or 24-bit colors. I am not in favor of this option.

Right, i forgot about the return values... indeed that will make things even more confusing.

> Alternative 1 is the one I think would be most elegant from a user POV, if the building blocks were there. I feel it ought to be a non-managed struct with only one field, but you'd need to be able to pass it to as an argument. It's probably not so much better that it justifies the work.

I *suppose* that if we get plain struct copied by value someday in the future, then we could also replace xColor properties type to Color struct, which will could have a constructor and/or factory methods. Maybe it would be possible even to cast integer value to Color struct directly, implicitly calling its initializer (hence achieving script compatibility). I'm only imagining possibilities right now.

> Aside from that, I would also suggest replacing COLOR_TRANSPARENT with Colors.Transparent, a pseudo-enum that would allow a list of basic colors: Colors.Black, .White, .Red, .Green, .Blue, .Magenta, .Yellow, .Aqua, etc.

I don't know how to do this in AGS, currently it does not support such syntax for enums. Maybe this could be left for later, after it is supported.

The rest of API suggestions should rather be put into a separate task imo, in order to not overload this one.

**Comment by ericoporto** [comment]
I am tending back to the plain int approach, specially because we now have both explicitly 32-bit and 8-bit color dynamic sprites. I am back playing with my cool things and 8-bit is definitely much easier to use when using sprites to encode levels, maps and other game elements.

**Comment by ivan-mogilko** [comment]
After #2648 the color properties are true ARGB now, but at the moment they are "clamped" to RGB by the property grid in the editor, and they probably won't work or may cause unexpected effects if any alpha value is used when assigning them in script, besides 0 and 255.

**Comment by ericoporto** [comment]
I think 2 is done, but not 3 and 5 was split in it's own issue.

**Comment by ivan-mogilko** [comment]
So, what is left here, more or less, is -

1. Make sure that all colored things in game support alpha:
* I think that all texts should already support it (both raw drawn, and displayed as a speech etc), because font renderers now respect alpha automatically, but it's worth to double check that it is not lost somewhere in the process.
* GUIs that may have background, border and text colors (text colors probably work, see above).
* Screen fade (?).
* Do I miss anything?

2. Amend script api with helper functions.
* Amend Game.GetColorFromRGB with alpha parameter, or make another function that makes ARGB.
* Helpers for getting color components from a color number, like GetR, GetG, GetB, GetA.
* BTW, maybe we should add a new Color struct, and use it as a collection of static functions for now (maybe it can become a color object in some future).

**Comment by AlanDrake** [comment]
> * Amend Game.GetColorFromRGB with alpha parameter, or make another function that makes ARGB.

I'd say to make a new Game.GetColorFromRGBA() so they're self explanatory.

> * Helpers for getting color components from a color number, like GetR, GetG, GetB, GetA.
> 
> * BTW, maybe we should add a new Color struct, and use it as a collection of static functions for now (with a potential to become a color object in some future).

This could start as a module. I've been playing with the concept and could try building something complete so we can explore how it would work. How it would perform, and stuff.

**Comment by ivan-mogilko** [comment]
About having script module, I think that it's a good thing to have this as a engine api anyway, because this is related to the basic operations over color data. If performance is an issue, a person may always write their own script in an optimized way, but engine api will still be there for when performance is not a problem, or user does not have time to write extra scripts for that.

**Comment by ivan-mogilko** [comment]
Added GetColorFromRGBA: c9367d9f12e4e8d4b7bc534c513a4878006ab307

**Comment by ivan-mogilko** [comment]
I tested, and apparently all color properties may already be with alpha: gui and speech draws proper colors. Same is with the speech, and text windows. I could have missed some setting, but I suppose it will be easier to fix if someone notices if one does not work.

Enabled setting alpha in color properties in the Editor: f64d837ef2d7cd573b62ad845a58d6f697cad6dc
Alpha is added as the last component, and is optional, that's more convenient imho.
As a reminder, currently you can only type it manually, color picker does not have alpha selection (there's a separate ticket for that).

I shall close this ticket as complete now.
